java/java.hints/src/org/netbeans/modules/java/hints/jdk/CanUseVT.java

```diff
@@ -0,0 +1,67 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.hints.jdk;
+
+import org.netbeans.api.java.queries.CompilerOptionsQuery;
+import org.netbeans.modules.java.hints.errors.Utilities;
+import org.netbeans.spi.editor.hints.ErrorDescription;
+import org.netbeans.spi.editor.hints.Fix;
+import org.netbeans.spi.java.hints.ConstraintVariableType;
+import org.netbeans.spi.java.hints.ErrorDescriptionFactory;
+import org.netbeans.spi.java.hints.Hint;
+import org.netbeans.spi.java.hints.HintContext;
+import org.netbeans.spi.java.hints.TriggerPattern;
+import org.netbeans.spi.java.hints.TriggerPatterns;
+import org.openide.util.NbBundle;
+
+/**
+ *
+ * @author mjayan
+ */
+@NbBundle.Messages({
+    "DN_CanUseVT=Can use Virtual Threads",
+    "DESC_CanUseVT=Can use virtual threads if workload is not CPU bound or number of concurrent tasks is high",
+    "ERR_CanUseVT=Can Use Virtual Thread Executor",
+    "ERR_CanUseThreadPerTask=Can Use Executors.newThreadPerTaskExecutor"
+})
+@Hint(displayName = "#DN_CanUseVT", description = "#DESC_CanUseVT", category = "rules15",
+        minSourceVersion = "19")
+public class CanUseVT {
+
+    private static final int VT_PREVIEW_JDK_VERSION = 19;
+
+    @TriggerPatterns({
+        @TriggerPattern(value = "java.util.concurrent.Executors.newFixedThreadPool($size)", constraints = @ConstraintVariableType(variable = "$size", type = "int")),
+        @TriggerPattern(value = "java.util.concurrent.Executors.newFixedThreadPool($size, $factory)", constraints = {
+            @ConstraintVariableType(variable = "$factory", type = "java.util.concurrent.ThreadFactory"),
+            @ConstraintVariableType(variable = "$size", type = "int")}),
+        @TriggerPattern(value = "java.util.concurrent.Executors.newCachedThreadPool()"),
+        @TriggerPattern(value = "java.util.concurrent.Executors.newCachedThreadPool($factory)", constraints
+                = @ConstraintVariableType(variable = "$factory", type = "java.util.concurrent.ThreadFactory"))})
+    public static ErrorDescription compute(HintContext ctx) {
+        if (Utilities.isJDKVersionLower(VT_PREVIEW_JDK_VERSION) && !CompilerOptionsQuery.getOptions(ctx.getInfo().getFileObject()).getArguments().contains("--enable-preview")) {
+            return null;
+        }
+        if (ctx.getVariables().get("$factory") != null) {
+            return ErrorDescriptionFactory.forName(ctx, ctx.getPath(), Bundle.ERR_CanUseThreadPerTask(), new Fix[0]);
+        } else {
+            return ErrorDescriptionFactory.forName(ctx, ctx.getPath(), Bundle.ERR_CanUseVT(), new Fix[0]);
+        }
```

Comment:
this could return a rewrite fix, otherwise it will just mark the line and do nothing.
something like:

```java
Fix fix = JavaFixUtilities.rewriteFix(ctx, "converts to VT msg", ctx.getPath(), "java.util.concurrent.Executors.newThreadPerTaskExecutor($factory)");
```
and...
```java
Fix fix = JavaFixUtilities.rewriteFix(ctx, "converts to VT msg", ctx.getPath(), "java.util.concurrent.Executors.newVirtualThreadPerTaskExecutor()");
```

Reply:
I understand your point. One of the reasons why we decided to give the hint was so that user is notified of  the support of virtual threads and he/she can take a call on whether to modify the current code to use virtual threads depending on the use case. This is why we are providing just the suggestion without rewriting the code.

In the initial commit there was rewrite logic as well, but after the first round of discussion, it was decided just to go with hints and not rewrite anything as there could be multiple corner cases like Executors.newFixedThreadPool(1) [as mentioned by @lahodaj  [here](https://github.com/apache/netbeans/pull/4592#pullrequestreview-1112236691)] which probably should not be rewritten.
This is the background. Please let me know your thoughts. If you think its best not to provide hints, I can update accordingly.

Reply:
sorry for the delay.

I believe this could be definitively useful as refactor-and-transform inspection. A dev can inspect a project and check if there are pools which can be moved to VTs on a case by case basis. If the code rule can convert the easy cases to the VT executor - why not (the difficult cases would be only marked without a transformation).

I just don't think it should be enabled by default, because light bulbs can be perceived as something you should fix. But if a dev enables the hint manually, the situation is different since the choice has been already made.

Those are just my thoughts, I don't have a strong opinion on it.

Reply:
@mbien : I have updated the code after discussion with @lahodaj. Modified the category to "suggestions" and  hintKind to ACTION in the latest code so that light bulbs are not shown by default whenever the thread pools are used. Havent implemented the conversion for easy cases as we decided against it due to the corner case [there are probably patterns that should not be re-written, like Executors.newFixedThreadPool(1)] Please check and let me know if this is ok.

Reply:
what do you think about `hintKind = Hint.Kind.INSPECTION, severity = Severity.HINT`, would this be also ok?

advantage:
 - it can be used as inspection too
 - it remains current-line-only just like with `hintKind = ACTION`

actions won't be listed in the source->inspect and refactor->inspect and transform dialogs and the user can't change the setting from current-line to warning since the combo box is disabled.

Reply:
@mbien : I think it should be ok to modify the hintKind to INSPECTION. It will be listed in source-> inspect. But as per the previous conclusion we dont provide any fix for conversion to virtual threads. So refactor-> inspect and transform will not work for transformation.

Reply:
@MegJayan right, there would be no transformation offered if its used via inspect and transform. But it would be current-line-only just like an action (but also configurable by the user) with the added benefit of it being usable as an inspection too (which I imagine could be super useful in some situations).

I don't know how to hide hints from the transformation dialog if they don't offer fixes. Maybe a third kind `{ACTION, INSPECTION, TRANSFORMATION}` would be required for that?

Reply:
I have updated the PR changing hintKind to INSPECTION. Since transformation is not available, it shows the below msg on clicking Inspect and Transform:
No fix for: Can Use Virtual Thread Executor at JavaApplication2.java:21
I think this conveys to the user that transformation isnt available and this should be ok.

Reply:
thank you @MegJayan for your patience. I am looking forward to use the new inspection/hint to find candidates for VTs.